README.md
# How to run the Ceylon Herd Module Repository

Note: at the moment Herd requires Play 1.2 which does not work on Java 8, so proceed with Java 7. We
are going to move it to Play 1.3 as soon as possible to fix that.

1. [Download Play Framework 1.2.7.2](https://downloads.typesafe.com/play/1.2.7.2/play-1.2.7.2.zip) and install it
1. Clone the [Herd repository](https://github.com/ceylon/ceylon-herd)
1. Open a shell and go to the `ceylon-herd` directory you just cloned
1. Run `play dependencies` to setup the required modules
    - Note: the `play` script requires Python 2, but runs on `/usr/bin/env python`,
      which may default to Python 3 depending on your system.
      If you get an error like this:

      ```
        File "/tmp/play-1.2.7.2/play", line 50
          print r"~        _            _ "
                                          ^
      SyntaxError: Missing parentheses in call to 'print'
      ```
      then change the first line of the script to
      ```python
      #!/usr/bin/env python2
      ```
    - Note 2: This will pretend to fail due to not being able to find `db` but
      in fact they are already there. This is due to Play apparently having removed their Maven
      repo online. Those two modules have been added to our Git repo so we have them, but this
      step is still necessary in order to set up the `secure` module which comes from the Play
      distribution itself. Ignore the two errors and move on: it should work.
1. Create your Postgres DB
    1. `sudo su - postgres`
    1. `createuser -PSRD ceylon-herd`
    1. _enter `ceylon-herd` as password when prompted_
    1. `createdb -O ceylon-herd -E utf8 ceylon-herd`
    1. exit
1. Run the application
    1. `play run`

# How to make your user (if registration is disabled, which is the default for now)

This can only be done by hand for now:

1. Pick a (dummy temporary) password
1. Hash it with BCrypt (for example, online at http://bcrypthashgenerator.apphb.com or other locations)
    1. Copy the resulting BCrypt Hash
    1. Don't worry about giving your password online, just pick a dummy temporary password and you can
       change it later in Herd.
1. Open a `psql` console to your database:
    1. `psql -h localhost -U ceylon-herd`
1. Add your user (as admin)
    1. `INSERT INTO user_table  (id, email, firstname, admin, lastname, isbcrypt, password, status, username) VALUES ((select nextval('hibernate_sequence')), 'email@example.org', 'FirstName', true, 'LastName', true, 'YOUR_BCRYPT_HASH', 'REGISTERED', 'UserName');`
1. You can now log in and change your password using the UI

# How to make your user admin (if registration is enabled)

This can only be done by hand for now:

1. Register your user at http://localhost:9000/register
1. See in the logs in the console what your activation link is, and follow it (in DEV mode no mail is sent)
1. Complete your registration
1. Open a `psql` console to your database:
    1. `psql -h localhost -U ceylon-herd`
1. Set yourself as admin
    1. `UPDATE user_table SET admin = true WHERE username = 'your-user-name';`

# Updating the DB to a new version

Just run the `query-db-version.sh` command to see if and how to update

# Troubleshooting

1. If you get the error `FATAL: Ident authentication failed for user "ceylon-herd"` look [here for a possible solution](http://www.cyberciti.biz/faq/psql-fatal-ident-authentication-failed-for-user/)

# Setting up Herd in Eclipse

1. Create a project using Play by going to the `ceylon-herd` project and running:
    1. `play eclipsify`
1. Import the resulting project into Eclipse

This also creates some useful launch configurations in the `eclipse`
folder that you can use to run the application within Eclipse.
But unfortunately the configurations are not 100% correct, so to make
things work you need to go into the properties of each configuration and
add the following to the `VM Arguments` section: `-XX:-UseSplitVerifier -Dfile.encoding=utf-8 -XX:CompileCommand=exclude,jregex/Pretokenizer,next`
    
# License

The content of this repository is released under ASL2 as provided in
the LICENSE file that accompanied this code.



debian/changelog
ceylon-herd (1.6) oneiric; urgency=low

  * New version

 -- Stephane Epardaud <stephane@epardaud.fr>  Wed, 11 Apr 2012 22:47:31 +0200

ceylon-herd (1.5) oneiric; urgency=low

  * New version

 -- Stephane Epardaud <stephane@epardaud.fr>  Tue, 10 Apr 2012 14:08:45 +0200

ceylon-herd (1.4) oneiric; urgency=low

  * New version

 -- Stephane Epardaud <stephane@epardaud.fr>  Tue, 03 Apr 2012 17:25:55 +0200

ceylon-herd (1.3) oneiric; urgency=low

  * Added admins and new identity

 -- Stephane Epardaud <stephane@epardaud.fr>  Fri, 02 Mar 2012 15:45:00 +0100

ceylon-herd (1.2) oneiric; urgency=low

  * New DB
  * Added stuff

 -- Stephane Epardaud <stephane@epardaud.fr>  Wed, 29 Feb 2012 00:38:12 +0100

ceylon-herd (1.0) UNRELEASED; urgency=low

  * Initial release

 -- Stephane Epardaud <stephane@epardaud.fr>  Sun, 26 Feb 2012 22:23:26 +0100


app/Bootstrap.java
/********************************************************************************
 * Copyright (c) 2011-2017 Red Hat Inc. and/or its affiliates and others
 *
 * This program and the accompanying materials are made available under the 
 * terms of the Apache License, Version 2.0 which is available at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: Apache-2.0 
 ********************************************************************************/
import java.io.File;
import java.util.EnumSet;

import models.Author;
import models.Dependency;
import models.HerdMetainf;
import models.Module;
import models.ModuleVersion;
import models.User;
import models.UserStatus;
import play.Play;
import play.jobs.Job;
import play.jobs.OnApplicationStart;
import util.CeylonElementType;
import util.Util;
 
@OnApplicationStart
public class Bootstrap extends Job {
 
    private static final String ASL = "Apache Software License";
    private static final String ASL2 = "Apache Software License 2.0";

    public void doJob() {
       
    	File repo = Util.getRepoDir();
    	if (!repo.exists()) {
    		repo.mkdir();
    	}
    	
    	File uploads = Util.getUploadDir();
    	if (!uploads.exists()) {
    		uploads.mkdir();
    	}
    	
    	if (Play.mode.isDev()) {
    	    HerdMetainf.initialize();
    	    if(Play.runingInTestMode()){
    	        loadTestDb();
    	    }
    	}
    	
    }

    private void loadTestDb() {
        User owner = new User();
        owner.email = "stef@epardaud.fr";
        owner.firstName = "Stef";
        owner.lastName = "Epardaud";
        owner.isBCrypt = true;
        owner.userName = "FroMage";
        owner.status = UserStatus.REGISTERED;
        owner.create();
        
        Module frEpardaudTest = createModule("fr.epardaud.test", owner);
        Module frEpardaudTest2 = createModule("fr.epardaud.test2", owner);
        Module frEpardaudJava = createModule("fr.epardaud.java", owner);
        Module ceylonCollection = createModule("ceylon.collection", owner);
        
        Author stef = createAuthor("Stéphane Épardaud");
        Author tom = createAuthor("Tom Bentley");
        Author enrique = createAuthor("Enrique Zamudio");

        ModuleVersion ceylonCollection666 = newModuleVersion(ceylonCollection, "6.6.6", ASL2, "Collection documentation", stef, tom, enrique);
        setArtifacts(ceylonCollection666, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 9);
        ceylonCollection666.create();

        // fr.epardaud.test
        
        ModuleVersion frEpardaudTest12 = newModuleVersion(frEpardaudTest, "1.2", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest12, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 9);
        frEpardaudTest12.create();
        addDependency(frEpardaudTest12, ceylonCollection666, false, true);

        ModuleVersion frEpardaudTest13 = newModuleVersion(frEpardaudTest, "1.3", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest13, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 7, 7);
        frEpardaudTest13.create();

        ModuleVersion frEpardaudTest14 = newModuleVersion(frEpardaudTest, "1.4", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest14, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 8);
        frEpardaudTest14.create();

        ModuleVersion frEpardaudTest15 = newModuleVersion(frEpardaudTest, "1.5", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest15, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 9, 8);
        frEpardaudTest15.create();

        ModuleVersion frEpardaudTest16 = newModuleVersion(frEpardaudTest, "1.6", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest16, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 9, 9);
        frEpardaudTest16.create();

        ModuleVersion frEpardaudTest17 = newModuleVersion(frEpardaudTest, "1.7", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest17, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 10, 10);
        frEpardaudTest17.create();

        ModuleVersion frEpardaudTest23 = newModuleVersion(frEpardaudTest, "2.3", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest23, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 9);
        frEpardaudTest23.create();
        addDependency(frEpardaudTest23, ceylonCollection666, false, true);
        frEpardaudTest23.addMember("fr.epardaud.test", "ParseException", CeylonElementType.Class, true);

        // fr.epardaud.test2

        ModuleVersion frEpardaudTest2_10 = newModuleVersion(frEpardaudTest2, "1.0", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest2_10, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 9);
        frEpardaudTest2_10.create();
        addDependency(frEpardaudTest2_10, ceylonCollection666, false, true);

        ModuleVersion frEpardaudTest2_12 = newModuleVersion(frEpardaudTest2, "1.2", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest2_12, EnumSet.of(Artifact.CAR, Artifact.JS, Artifact.SRC), 8, 9);
        frEpardaudTest2_12.create();
        addDependency(frEpardaudTest2_12, ceylonCollection666, false, true);
        frEpardaudTest2_12.addMember("fr.epardaud.test2", "ParseException", CeylonElementType.Class, true);
        frEpardaudTest2_12.addMember("fr.epardaud.test2.float", "pi", CeylonElementType.Value, true);

        // only JVM
        ModuleVersion frEpardaudTest03 = newModuleVersion(frEpardaudTest2, "0.3", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest03, EnumSet.of(Artifact.CAR, Artifact.SRC), 8, 0);
        frEpardaudTest03.create();
        addDependency(frEpardaudTest03, ceylonCollection666, false, true);

        // only JS
        ModuleVersion frEpardaudTest04 = newModuleVersion(frEpardaudTest2, "0.4", ASL, "Super documentation", stef, tom);
        setArtifacts(frEpardaudTest04, EnumSet.of(Artifact.JS, Artifact.SRC), 0, 9);
        frEpardaudTest04.create();
        addDependency(frEpardaudTest04, ceylonCollection666, false, true);

        // fr.epardaud.java
        
        ModuleVersion frEpardaudJava1 = newModuleVersion(frEpardaudJava, "1", null, null);
        setArtifacts(frEpardaudJava1, EnumSet.of(Artifact.JAR), 0, 0);
        frEpardaudJava1.create();
    }

    private enum Artifact {
        CAR, JAR, JS, SRC; 
    }
    
    private void setArtifacts(ModuleVersion mv, EnumSet<Artifact> artifacts, 
            int jvmBinaryMajor, int jsBinaryMajor) {
        mv.isCarPresent = artifacts.contains(Artifact.CAR);
        mv.isJsPresent = artifacts.contains(Artifact.JS);
        mv.isJarPresent = artifacts.contains(Artifact.JAR);
        mv.isSourcePresent = artifacts.contains(Artifact.SRC);
        mv.jvmBinMajor = jvmBinaryMajor;
        mv.jsBinMajor = jsBinaryMajor;
    }

    private void addDependency(ModuleVersion owner, ModuleVersion dependency, boolean optional, boolean shared) {
        Dependency dep = new Dependency(owner, dependency.module.name, dependency.version, 
                optional, shared, false, false, false, false);
        dep.create();
    }

    private ModuleVersion newModuleVersion(Module module, String version, String license, String doc, Author... authors) {
        ModuleVersion mv = new ModuleVersion();
        mv.module = module;
        mv.version = version;
        mv.license = license;
        mv.doc = doc;
        for(Author author : authors)
            mv.authors.add(author);
        return mv;
    }

    private Author createAuthor(String name) {
        Author author = new Author();
        author.name = name;
        author.create();
        return author;
    }

    private Module createModule(String name, User owner) {
        Module module = new Module();
        module.name = name;
        module.owner = owner;
        module.create();
        return module;
    }
 
}